### PR TITLE
Add guidance for debug packages

### LINKED ISSUE #3697: Provide some guidelines to package pdbs.
Prior to .NET 8, the pdbs that were created as part of the build weren't usable because they referenced the source files that were used during the distro packaging.

.NET 8 produces usable pdbs thanks to source-link (which makes pdbs reference the code at https://github.com/dotnet/dotnet).

The pdbs get bundled in a tarfile that is output by the vmr build.
The tarball matches the layout of the .NET installation.
For example, this is the contents of a this tarball for an 8.0 build:

```
584K	packs/Microsoft.AspNetCore.App.Ref
12M	packs/Microsoft.AspNetCore.App.Runtime.fedora.37-x64
1.3M	packs/Microsoft.NETCore.App.Ref
21M	packs/Microsoft.NETCore.App.Runtime.fedora.37-x64
92M	sdk/8.0.100-rtm.23527.1
12M	shared/Microsoft.AspNetCore.App
21M	shared/Microsoft.NETCore.App
```

We should add some guidelines to [packaging guideline](https://learn.microsoft.com/en-us/dotnet/core/distribution-packaging) that describe how to package these files.

cc @MichaelSimons @omajid @dotnet/distro-maintainers

### PR DESCRIPTION
**Pull request by NikolaMilosavljevic**
Fixes https://github.com/dotnet/source-build/issues/3697

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/distribution-packaging.md](https://github.com/dotnet/docs/blob/ab2173e09ede6e8182126353865896e15b7bbc66/docs/core/distribution-packaging.md) | [.NET distribution packaging](https://review.learn.microsoft.com/en-us/dotnet/core/distribution-packaging?branch=pr-en-us-38546) |

### CODE REVIEW

File: docs/core/distribution-packaging.md [hunk 1]
```diff
@@ -198,6 +198,55 @@ Most distributions require all artifacts to be built from source. This has some
...
+  - **Contains:** debug content for (6)
+  - **Dependencies:** `aspnetcore-runtime-[major].[minor]`
+
+- `dotnet-runtime-dbg-[major].[minor]` - Installs debug content for a specific runtime
+  - **Version:** \<runtime version>
+  - **Example:** dotnet-runtime-dbg-8.0
+  - **Contains:** debug content for (5)
+  - **Dependencies:** `dotnet-runtime-[major].[minor]`
+
+Debug tarball also contains some debug content under `packs` which represent copies of content under `shared`. In .NET layout, `packs` directory is used for building .NET applications - there are no debugging scenarios. Thus debug content under `packs`, in debug tarball, should not be packaged.
```
Comment (MichaelSimons) [inline docs/core/distribution-packaging.md:248]:
```suggestion
The debug tarball also contains some debug content under `packs` which represent copies of content under `shared`. In the .NET layout, the `packs` directory is used for building .NET applications - there are no debugging scenarios, therefore the debug content under `packs`, in the debug tarball, should not be packaged.
```

Comment (MichaelSimons) [inline docs/core/distribution-packaging.md:248]: Would it make sense for use to strip the packs directory from the sdk debug tarball?

Comment (NikolaMilosavljevic) [reply docs/core/distribution-packaging.md]: Yes, that is probably the best option going forward. I've created the issue: https://github.com/dotnet/source-build/issues/3769

We'll fix it in 9.0 and once backported to 8.0, we can update the guidance.

**Review by tmds — APPROVED** [review]

**Comment by tmds** [comment]
LGTM, thanks for writing this down!

File: docs/core/distribution-packaging.md [hunk 2]
```diff
@@ -198,6 +198,55 @@ Most distributions require all artifacts to be built from source. This has some
...
+
+Starting with .NET 8.0, all .NET debug content (PDB files), produced by source-build, is available in a tarball named `dotnet-symbols-sdk-<version>-<rid>.tar.gz`. This archive contains PDBs in sub-directories that match directory structure of .NET SDK tarball - `dotnet-symbols-<version>-<rid>.tar.gz`.
+
+While all debug content is available in debug tarball, not all debug content is equally important. End users are mostly interested in content of `shared/Microsoft.AspNetCore.App/<aspnetcore version>` and `shared/Microsoft.NETCore.App/<runtime version>`.
+
+SDK content, under `sdk/<sdk version>` is useful for debugging of .NET SDK toolsets.
+
+The following lists the recommended debug packages:
+
+- `dotnet-sdk-dbg-[major].[minor]` - Installs debug content for a specific SDK version
```
[the remark this answers is not in the document]
Comment (NikolaMilosavljevic) [reply docs/core/distribution-packaging.md]: Fixed with https://github.com/dotnet/docs/pull/38546/commits/376550fd3ed30555b3b302fa08c8ee0675a134b9

**Review by MichaelSimons — APPROVED** [review]

File: docs/core/distribution-packaging.md [hunk 3]
```diff
@@ -198,6 +198,57 @@ Most distributions require all artifacts to be built from source. This has some
 
 Multiple `dotnet-sdk` packages may provide the same files for the `NuGetFallbackFolder`. To avoid issues with the package manager, these files should be identical (checksum, modification date, and so on).
 
+### Debug packages
+
+Debug content should be packaged in debug-named packages that follow the .NET package split described above. For instance, debug content for `dotnet-sdk-[major].[minor]` package should be included in a package named `dotnet-sdk-dbg-[major].[minor]`. Debug content should be installed to the same location as the binaries.
```
Comment (gewarren) [inline docs/core/distribution-packaging.md:203]:
```suggestion
Debug content should be packaged in debug-named packages that follow the .NET package split described previously in this article. For instance, debug content for the `dotnet-sdk-[major].[minor]` package should be included in a package named `dotnet-sdk-dbg-[major].[minor]`. You should install debug content to the same location as the binaries.
```

File: docs/core/distribution-packaging.md [hunk 4]
```diff
@@ -198,6 +198,57 @@ Most distributions require all artifacts to be built from source. This has some
...
+
+Few binary examples:
```
Comment (gewarren) [inline docs/core/distribution-packaging.md:205]:
```suggestion
Here are a few binary examples:
```

File: docs/core/distribution-packaging.md [hunk 5]
```diff
@@ -198,6 +198,57 @@ Most distributions require all artifacts to be built from source. This has some
...
+
+In `{dotnet_root}/sdk/<sdk version>` directory, we'd expect the following two files:
```
Comment (gewarren) [inline docs/core/distribution-packaging.md:207]:
```suggestion
In the `{dotnet_root}/sdk/<sdk version>` directory, the following two files are expected:
```

File: docs/core/distribution-packaging.md [hunk 6]
```diff
@@ -198,6 +198,57 @@ Most distributions require all artifacts to be built from source. This has some
...
+
+- `dotnet.dll` - installed with `dotnet-sdk-[major].[minor]` package
+- `dotnet.pdb` - installed with `dotnet-sdk-dbg-[major].[minor]` package
+
+In `{dotnet_root}/shared/Microsoft.NETCore.App/<runtime version>` we'd expect the following two files:
```
Comment (gewarren) [inline docs/core/distribution-packaging.md:212]:
```suggestion
In the `{dotnet_root}/shared/Microsoft.NETCore.App/<runtime version>` directory, the following two files are expected:
```

File: docs/core/distribution-packaging.md [hunk 7]
```diff
@@ -198,6 +198,57 @@ Most distributions require all artifacts to be built from source. This has some
...
+
+- `System.Text.Json.dll` - installed with `dotnet-runtime-[major].[minor]` package
+- `System.Text.Json.pdb` - installed with `dotnet-runtime-dbg-[major].[minor]` package
+
+In `{dotnet_root/shared/Microsoft.AspNetCore.App/<aspnetcore version>` we'd expect the following two files:
```
Comment (gewarren) [inline docs/core/distribution-packaging.md:217]:
```suggestion
In the `{dotnet_root/shared/Microsoft.AspNetCore.App/<aspnetcore version>` directory, the following two files are expected:
```

File: docs/core/distribution-packaging.md [hunk 8]
```diff
@@ -198,6 +198,57 @@ Most distributions require all artifacts to be built from source. This has some
...
+
+- `Microsoft.AspNetCore.Routing.dll` - installed with `aspnetcore-runtime-[major].[minor]` packages
+- `Microsoft.AspNetCore.Routing.pdb` - installed with `aspnetcore-runtime-dbg-[major].[minor]` packages
+
+Starting with .NET 8.0, all .NET debug content (PDB files), produced by source-build, is available in a tarball named `dotnet-symbols-sdk-<version>-<rid>.tar.gz`. This archive contains PDBs in sub-directories that match directory structure of .NET SDK tarball - `dotnet-symbols-<version>-<rid>.tar.gz`.
```
Comment (gewarren) [inline docs/core/distribution-packaging.md:222]:
```suggestion
Starting with .NET 8.0, all .NET debug content (PDB files), produced by source-build, is available in a tarball named `dotnet-symbols-sdk-<version>-<rid>.tar.gz`. This archive contains PDBs in subdirectories that match the directory structure of the .NET SDK tarball - `dotnet-symbols-<version>-<rid>.tar.gz`.
```

File: docs/core/distribution-packaging.md [hunk 9]
```diff
@@ -198,6 +198,57 @@ Most distributions require all artifacts to be built from source. This has some
...
+
+While all debug content is available in debug tarball, not all debug content is equally important. End users are mostly interested in content of `shared/Microsoft.AspNetCore.App/<aspnetcore version>` and `shared/Microsoft.NETCore.App/<runtime version>`.
```
Comment (gewarren) [inline docs/core/distribution-packaging.md:224]:
```suggestion
While all debug content is available in the debug tarball, not all debug content is equally important. End users are mostly interested in the content of the `shared/Microsoft.AspNetCore.App/<aspnetcore version>` and `shared/Microsoft.NETCore.App/<runtime version>` directories.
```

File: docs/core/distribution-packaging.md [hunk 10]
```diff
@@ -198,6 +198,57 @@ Most distributions require all artifacts to be built from source. This has some
...
+
+SDK content, under `sdk/<sdk version>` is useful for debugging of .NET SDK toolsets.
```
Comment (gewarren) [inline docs/core/distribution-packaging.md:226]:
```suggestion
The SDK content under `sdk/<sdk version>` is useful for debugging .NET SDK toolsets.
```

File: docs/core/distribution-packaging.md [hunk 11]
```diff
@@ -198,6 +198,57 @@ Most distributions require all artifacts to be built from source. This has some
...
+
+The following lists the recommended debug packages:
```
Comment (gewarren) [inline docs/core/distribution-packaging.md:228]:
```suggestion
The following packages are the recommended debug packages:
```

File: docs/core/distribution-packaging.md [hunk 12]
```diff
@@ -198,6 +198,57 @@ Most distributions require all artifacts to be built from source. This has some
...
+
+The following debug package is optional:
+
+- `dotnet-sdk-dbg-[major].[minor]` - Installs debug content for a specific SDK version
+  - **Version:** \<sdk version>
+  - **Example:** dotnet-sdk-dbg-8.0
+  - **Contains:** debug content for (3),(4),(18)
+  - **Dependencies:** `dotnet-sdk-[major].[minor]`
+
+The debug tarball also contains some debug content under `packs` which represent copies of content under `shared`. In the .NET layout, the `packs` directory is used for building .NET applications - there are no debugging scenarios, therefore the debug content under `packs`, in the debug tarball, should not be packaged.
```
Comment (gewarren) [inline docs/core/distribution-packaging.md:250]:
```suggestion
The debug tarball also contains some debug content under `packs`, which represents copies of content under `shared`. In the .NET layout, the `packs` directory is used for *building* .NET applications. There are no debugging scenarios, so you shouldn't package the debug content under `packs` in the debug tarball.
```

**Review by gewarren — APPROVED** [review]

File: docs/core/distribution-packaging.md [hunk 13]
```diff
@@ -198,6 +198,57 @@ Most distributions require all artifacts to be built from source. This has some
 
 Multiple `dotnet-sdk` packages may provide the same files for the `NuGetFallbackFolder`. To avoid issues with the package manager, these files should be identical (checksum, modification date, and so on).
 
+### Debug packages
+
+Debug content should be packaged in debug-named packages that follow the .NET package split described previously in this article. For instance, debug content for the `dotnet-sdk-[major].[minor]` package should be included in a package named `dotnet-sdk-dbg-[major].[minor]`. You should install debug content to the same location as the binaries.
```
Comment (omajid) [inline docs/core/distribution-packaging.md:203]: Any thoughts on making this more flexible to comply with additional distro conventions? Fedora, for example, seems to prefer `debug` over `dbg`:

```
$ podman run -it fedora:39
# dnf list -q '*debug*' | wc -l
219
# dnf list -q '*dbg*' | wc -l
4
```

Comment (tmds) [reply docs/core/distribution-packaging.md:203]: Part of the value of these guidelines is for different distros to use the same names.

These are all "recommendations". A maintainer can still chose to do something different.

**Pull request merged**